system-contracts/scripts/compile-yul.ts

```diff
@@ -30,14 +42,34 @@ async function main() {
   program.version("0.1.0").name("compile yul").description("publish preimages for the L2 contracts");
 
   program.command("compile-bootloader").action(async () => {
-    await compileYulFolder("bootloader/build");
-    await compileYulFolder("bootloader/tests");
+    const timestampFilePath = _path.join(process.cwd(), TIMESTAMP_FILE_BOOTLOADER);
+    const folderToCheck = _path.join(process.cwd(), BOOTLOADER_DIR);
+
+    if (needsRecompilation(folderToCheck, timestampFilePath)) {
+      console.log("Compilation needed.");
+      await compileYulFolder("bootloader/build");
+      await compileYulFolder("bootloader/tests");
+      setCompilationTime(timestampFilePath);
+    } else {
+      console.log("Compilation not needed.");
+      return;
+    }
   });
 
   program.command("compile-precompiles").action(async () => {
-    await compileYulFolder("contracts-preprocessed");
-    await compileYulFolder("contracts-preprocessed/precompiles");
-    await compileYulFolder("contracts-preprocessed/precompiles/test-contracts");
+    const timestampFilePath = _path.join(process.cwd(), TIMESTAMP_FILE_YUL);
+    const folderToCheck = _path.join(process.cwd(), CONTRACTS_DIR);
+
+    if (needsRecompilation(folderToCheck, timestampFilePath)) {
```

Comment:
Checking the folder, rather than specific subfolders, like "contracts-preprocessed/precompiles/test-contracts". 
If necessary, we can check each of the compilation targets separately